File: tree/94.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
    s := make([]*TreeNode, 0)
    l := make([]int, 0)
    for {
        for root != nil {
            s = append(s, root)
            root = root.Left
        }
        if len(s) == 0 {
            break
        }
        root = s[len(s)-1]
        s = s[:len(s)-1]
        l = append(l, root.Val)
        root = root.Right
    }
    return l
}

File: tree/94.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        std::vector<int> ret;
        if (!root)
            return ret;
        std::stack<TreeNode*> s;
        while (true) {
            while (root) {
                s.emplace(root);
                root = root->left;
            }
            if (s.empty())
                break;
            root = s.top();
            s.pop();
            ret.emplace_back(root->val);
            root = root->right;
        }
        return ret;
    }
};
